### PR TITLE
Bug 1807237 - Optimize task that caches external gradle dependencies

### PR DESCRIPTION
**Pull request by JohanLorenzo**
This PR splits the task called `toolchain-linux64-android-gradle-dependencies` into a hundred of them. We now cache external gradle dependencies per android-component/apk. This makes things better in multiple aspects:

 1. We don't have a single big task anymore which takes an hour to complete. The longest task now takes 20 minutes[1]
 2. If one of these tasks intermittently fails, then Taskcluster automatically and solely reruns the one that failed - meaning it just adds 20 minutes at most.
 3. The rest of the graph of continues when the corresponding cache task finishes. This means we get some early feedback in case something goes wrong.
 4. The shell script is now smaller and less prone to bugs.

@csadilek I believe this is a better alternative to hosting a nexus instance because of traceability and monitoring. 

Suggested followup improvements:
 1. Schedule downstream tasks when whenever taskgraph spins a cache task (this wasn't supported before this PR either)
 2. Let the cache task check that no empty file was downloaded and if so, rerun the whole task 

As usual, I recommend reviewing each commit individually. 

[1] e.g.: https://firefox-ci-tc.services.mozilla.com/tasks/groups/RJAFpBC9QQCjj1YTAXIFcQ#external
https://bugzilla.mozilla.org/show_bug.cgi?id=1807237
https://bugzilla.mozilla.org/show_bug.cgi?id=1807237
https://bugzilla.mozilla.org/show_bug.cgi?id=1807237

### CODE REVIEW

**Review by csadilek — APPROVED** [review]
Thanks, this looks like a big improvement, and right now is the best time to test this more!

One thing I don't see / understand is how the cache that is created and zipped by the individual tasks is ultimately used when the toolchain task is not needed. Can you point me to that?

File: taskcluster/android_taskgraph/loader/build_config.py [hunk 1]
```diff
@@ -35,8 +48,20 @@ def components_loader(kind, path, config, params, loaded_tasks):
...
             and (component['shouldPublish'] or build_type == 'regular')
+            and (for_build_type is None or build_type == for_build_type)
         )
     }
-    config['tasks'] = tasks
 
-    return base_loader(kind, path, config, params, loaded_tasks)
+    return tasks
+
+def _get_apks_tasks(config):
```
[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply taskcluster/android_taskgraph/loader/build_config.py:57]: You're right, the naming isn't great and now that I look at it again, there's a code smell in here. I'll get that fixed. Thanks for spotting this one out!

**Comment by JohanLorenzo** [comment]
> One thing I don't see / understand is how the cache that is created and zipped by the individual tasks is ultimately used when the toolchain task is not needed. Can you point me to that?

Thanks for asking! The `toolchain` task actually got rebranded into `external-gradle-depedencies` and got split into 100 tasks. This means the principle remains the same: 

 1. One `external-gradle-dependencies` task `assemble` and `test` a given component
 2. The `pom`/`jar`/`aar` files are captured by a nexus proxy that runs within the task and gets killed afterwards. The proxy isn't shared among tasks.
 3. Like you said, these `pom`/`jar`/`aar` files are then zipped and exposed as a Taskcluster artifact
 4. The following tasks will download this Taskcluster artifact, extract the archive, and tell gradle to use these local files instead of maven central or Google's maven repo.

What really changes is: instead of having a big single archive shared among all tasks of all components, we now have a 100+ smaller ones. For the same commit, 2 different components will use 2 different archives that contain just the `pom`/`jar`/`aar` files for the given component. 

Let me know if you need more information 🙂

File: taskcluster/android_taskgraph/build_config.py [hunk 2]
```diff
@@ -71,6 +71,15 @@ def get_upstream_deps_for_all_gradle_projects():
     return all_deps
 
 
+def get_apks():
```
[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply taskcluster/android_taskgraph/build_config.py]: Renamed into `get_apk_based_projects()`. We don't need variants in this context. Let me know if you find a better name.

**Review by jcristau — APPROVED** [review]

File: taskcluster/android_taskgraph/loader/build_config.py [hunk 3]
```diff
@@ -4,11 +4,17 @@
 
 
 from taskgraph.loader.transform import loader as base_loader
-from taskgraph.util.templates import merge
 
 from ..build_config import get_components
 
 
+# Treeherder group are capped at 25 chars
```
[the remark this answers is not in the document]
Comment (JohanLorenzo) [reply taskcluster/android_taskgraph/loader/build_config.py]: Done! Minor side effect: the 2 exceptions now have an extra character at the end of the group: https://github.com/mozilla-mobile/firefox-android/pull/490/files#diff-78cef187ec6d54d1013f6c8b16706396955b3c4be208320c0fdf815418cce528R142

**Comment by JohanLorenzo** [comment]
@mergifyio backport releases_v109

**Comment by JohanLorenzo** [comment]
@Mergifyio backport releases_v110

**Pull request merged**